follow up #115078, broken distributed tests

jeffdaily opened the pull request:
ROCm distributed tests started failing after #115078.  This skips the new tests if the number of GPUs available isn't sufficient.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

jeffdaily committed:
follow up #115078, broken distributed tests

jeffdaily committed:
lint

jeffdaily commented:
@huydhn, @malfet, @atalman 

This PR passes the rocm periodic distributed shards 1 and 2.  Fixes current broken tests on HUD.  Please land.

malfet reviewed (changes requested):
This feels wrong, as `sys.exit` just abort test execution (and also not limited to ROCm)

Are you saying that `@skip_if_lt_x_gpu` decorator does not work correctly? Or does not work correctly for ROCm?

jeffdaily commented:
> This feels wrong, as `sys.exit` just abort test execution (and also not limited to ROCm)

Felt wrong for me, too, but that's what is done for distributed tests.  This logic is copied from the `@with_comms` decorator defined in this file, as well as from the `@skip_if_lt_x_gpu` from `torch.testing._internal.common_distributed`.

> Are you saying that `@skip_if_lt_x_gpu` decorator does not work correctly? Or does not work correctly for ROCm?

It works correctly.  But in this case, the `TestNCCLCollectivesWithWorldSize4` subclass of `TestCollectivesWithNCCL` inherits the tests from the base class.  It seems the intent of the current use of `@skip_if_lt_x_gpu(WORLD_SIZE)` isn't maintained for the subclass because the subclass changes its `self.world_size` but the decorators were using `WORLD_SIZE` and you can't directly use `self.world_size` within the decorator because `self` isn't available yet.  It really should be testing based on `self.world_size` and in fact the `@with_comms()` decorator is doing that.  In this PR I removed the superfluous use of `@skip_if_lt_x_gpu` in favor of `@with_comms()`.  Unfortunately, I could not use `@with_comms()` for the test `test_tracing_with_fakepg` because it is a fake PG, and that's why I added the `exit_if_lt_x_gpu` function.

jeffdaily commented:
@malfet  please re-review with my comments.

jithunnair-amd commented:
@malfet @huydhn We are trying to get ROCm distributed CI back to green, please help review and merge.

huydhn commented on test/distributed/test_functional_api.py:
```diff
@@ -427,6 +426,11 @@ def allred_mesh_dim(input):
 
+def exit_if_lt_x_gpu(x):
+    if BACKEND == dist.Backend.NCCL and torch.cuda.device_count() < x:
+        sys.exit(TEST_SKIPS[f"multi-gpu-{x}"].exit_code)
```
Is `sys.exit` to skip the test or exit the test suit completely?  As I see it is used in `test_tracing_with_fakepg`, I assume it would be the former (to skip that test).  Is that better to use https://docs.python.org/3/library/unittest.html#unittest.SkipTest instead of `sys.exit`

huydhn replied:
Oh, I now see @malfet comment above.

jeffdaily replied:
For distributed tests, sys.exit is used.  See the decorator `skip_if_lt_x_gpu` in `torch/testing/_internal/common_distributed.py`.

jeffdaily replied:
Can someone else experiment with not using sys.exit in a follow-up PR instead of having me do it here?  I'm trying to get our distributed CI signal back to green and I'm following the precedent established elsewhere in this file and elsewhere in torch/testing/_internal.

malfet replied:
Well, my problem with that decorator, is that we can skip a bunch of tests because we exit from the testsuite early?

As a middle ground can we check that exit code is non-zero here? Or should it be?

wconstab replied:
yea we shouldn't do exit here.  why is the skip_if_lt_x_gpu macro not working here?

jeffdaily replied:
See my previous comment https://github.com/pytorch/pytorch/pull/116217#issuecomment-1867012873.

jeffdaily commented:
@pytorchbot rebase

malfet reviewed (approved):
Ok, I see that `with_comms` decorates the same test as this one, but I'm still confused why this simply does not prevent a bunch of tests from running.

@rohan-varma , @wconstab do you know how this is supposed to work?

huydhn commented:
> Ok, I see that `with_comms` decorates the same test as this one, but I'm still confused why this simply does not prevent a bunch of tests from running.
> 
> @rohan-varma , @wconstab do you know how this is supposed to work?

IIRC, distributed tests are run with subprocess https://github.com/pytorch/pytorch/blob/main/test/run_test.py#L1092-L1106, so I guess `sys.exit` will only exit the child processes instead.

wanchaol commented:
> 
> IIRC, distributed tests are run with subprocess https://github.com/pytorch/pytorch/blob/main/test/run_test.py#L1092-L1106, so I guess `sys.exit` will only exit the child processes instead.

yes I think that's the expected behavior, most distributed test cases inherits from the MultiProcessTestCase, where before the actual test case run, we spawn multiple processes (including an additonal main process for monitoring iirc), so the `sys.exit` call only exit the child process @malfet

huydhn commented:
@jeffdaily Just FYI, I think there are still some broken ROCm distributed job leftover after this https://hud.pytorch.org/pytorch/pytorch/commit/db79ceb110f6646523019a59bbd7b838f43d4a86

jeffdaily commented:
> @jeffdaily Just FYI, I think there are still some leftover broken ROCm distributed tests after this https://hud.pytorch.org/pytorch/pytorch/commit/db79ceb110f6646523019a59bbd7b838f43d4a86

#116713 should fix the remaining distributed tests with its changes to torch/_inductor/codecache.py.
